activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -700,13 +699,10 @@ def checkout_for_exclusive_access(checkout_timeout)
         end
 
         def with_new_connections_blocked
-          previous_value = nil
-          synchronize do
-            previous_value, @new_cons_enabled = @new_cons_enabled, false
-          end
+          synchronize { @new_cons_blocks += 1 }
           yield
         ensure
-          synchronize { @new_cons_enabled = previous_value }
+          synchronize { @new_cons_blocks -= 1 }
```

Comment:
@thedarkone should we be servicing `@available` here, too? (if we just got to zero)

AFAICS, if a thread was blocked, it's now [waiting](https://github.com/kamipo/rails/blob/700639e4f6d4d9e9be50fe282756045f9189dd23/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L730), but no-one is arranging to make the connection for it -- it's just left to hope some other already-established connection will be returned in time.